### DISCUSSION TITLE
Decoding objects in qpdf json

### DISCUSSION THREAD

**Discussion by vulpes2** [qpdf/qpdf#1184] [Q&A]
I'm trying to remove some text watermarks in a PDF document to make it less annoying to read, and I had pretty good success with just qpdf alone (no pdftk). Unfortunately this is still somewhat of a manual process, as I had to open the decompressed PDF file with a text editor, search for the watermark strings, find the object number, and replace the base64 data for that object with an empty string in the entire json file. Sometimes this needs to be repeated multiple times to get rid of all watermarks in the document. 

Is it possible to decode the base64 data in the qpdf json to the same format that I would get by opening a PDF file written with `--stream-data=uncompress` so I could search for strings directly and automate this process?

**Comment by jberkenbilt** [comment]
If you use `--json-stream-data=file`, qpdf will write a separate file for each stream in a file next to the json. Those files are not encoded. You can use
```
qpdf  --json-output --json-stream-data=file --compress-streams=n --decode-level=generalized in.pdf /tmp/x/a.json
```
which will create /tmp/x/a.json-nn with the uncompressed contents of stream nn. You can directly edit those files and then use `qpdf --json-input /tmp/x/a.json /tmp/out.pdf` to reassemble the PDF with edited streams.

**Comment by jberkenbilt** [comment]
I would also refer you to my blog: https://medium.com/@jberkenbilt/examining-a-pdf-file-with-qpdf-219ff19d5c47 which discusses various other options including QDF mode and talks about this problem.
